File: ios/Sophia/Services/AnalyticsService.swift
import Foundation
import Mixpanel
import RevenueCat

enum AnalyticsService {
    private static let sessionGapSeconds: TimeInterval = 30 * 60
    private static let lastSessionKey = "sophia_analytics_last_session"
    private static let firstOpenKey = "sophia_analytics_first_open"

    static func configure() {
        let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown"
        Mixpanel.initialize(
            token: AppConfig.MIXPANEL_TOKEN,
            trackAutomaticEvents: false,
            superProperties: [
                "app_version": version,
                "platform": "ios",
            ],
            serverURL: "https://api-eu.mixpanel.com"
        )

        if UserDefaults.standard.object(forKey: firstOpenKey) == nil {
            UserDefaults.standard.set(Date().timeIntervalSince1970, forKey: firstOpenKey)
        }
    }

    static func updateUserContext(
        language: AppLanguage,
        isPremium: Bool,
        onboardingCompleted: Bool
    ) {
        let interests = Array(OnboardingViewModel.userInterestKeys()).sorted()
        let unlocked = Subject.allCases.map(\.storageKey).sorted()

        let mixpanel = Mixpanel.mainInstance()
        mixpanel.registerSuperProperties([
            "language": language.rawValue,
            "is_premium": isPremium,
            "onboarding_completed": onboardingCompleted,
            "unlocked_subjects": unlocked,
        ])
        mixpanel.people.set(properties: [
            "interests": interests,
            "is_premium": isPremium,
            "language": language.rawValue,
        ])
        linkRevenueCatUserIfNeeded()
    }

    static func linkRevenueCatUserIfNeeded() {
        guard Purchases.isConfigured else { return }
        let rcUserID = Purchases.shared.appUserID
        guard !rcUserID.isEmpty else { return }
        Mixpanel.mainInstance().identify(distinctId: rcUserID)
    }

    // MARK: - Lifecycle

    static func trackAppOpened(source: String) {
        track("app_opened", ["source": source])
    }

    static func trackSessionIfNeeded(isPremium: Bool) {
        let now = Date().timeIntervalSince1970
        let last = UserDefaults.standard.double(forKey: lastSessionKey)
        guard last == 0 || now - last >= sessionGapSeconds else { return }

        UserDefaults.standard.set(now, forKey: lastSessionKey)
        let firstOpen = UserDefaults.standard.double(forKey: firstOpenKey)
        let daysSinceInstall: Int
        if firstOpen > 0 {
            daysSinceInstall = max(0, Int((now - firstOpen) / 86_400))
        } else {
            daysSinceInstall = 0
        }

        track("session_started", [
            "days_since_install": daysSinceInstall,
            "is_premium": isPremium,
        ])
    }

    // MARK: - Onboarding

    static func trackOnboardingStarted() {
        track("onboarding_started")
    }

    static func trackOnboardingStepViewed(stepIndex: Int, action: String? = nil) {
        var props: [String: MixpanelType] = [
            "step_index": stepIndex,
            "step_name": onboardingStepName(for: stepIndex),
        ]
        if let action { props["action"] = action }
        track("onboarding_step_viewed", props)
    }

    static func trackOnboardingInterestsSet(_ interests: Set<String>) {
        track("onboarding_interests_set", [
            "interests": Array(interests).sorted(),
        ])
    }

    static func trackOnboardingCompleted(sawPaywall: Bool, isPremiumAtExit: Bool) {
        track("onboarding_completed", [
            "saw_paywall": sawPaywall,
            "is_premium_at_exit": isPremiumAtExit,
        ])
    }

    // MARK: - Paywall

    static func trackPaywallViewed(context: String, triggerCourseId: String? = nil) {
        var props: [String: MixpanelType] = ["context": context]
        if let triggerCourseId { props["trigger_course_id"] = triggerCourseId }
        track("paywall_viewed", props)
    }

    static func trackPaywallDismissed(context: String, durationSeconds: Int) {
        track("paywall_dismissed", [
            "context": context,
            "duration_seconds": durationSeconds,
        ])
    }

    static func trackDiscountOfferViewed(source: String) {
        track("discount_offer_viewed", ["source": source])
    }

    static func trackPurchaseCompleted(context: String, offeringId: String? = nil, packageId: String? = nil) {
        var props: [String: MixpanelType] = ["context": context]
        if let offeringId { props["offering_id"] = offeringId }
        if let packageId { props["package_id"] = packageId }
        track("purchase_completed", props)
        Mixpanel.mainInstance().people.set(property: "is_premium", to: true)
    }

    // MARK: - Course

    static func trackCourseOpened(
        courseId: String,
        subject: Subject,
        source: String,
        isFreeUser: Bool
    ) {
        track("course_opened", [
            "course_id": courseId,
            "subject": subject.storageKey,
            "source": source,
            "is_free_user": isFreeUser,
        ])
    }

    static func trackCourseCompleted(course: Course) {
        track("course_completed", [
            "course_id": course.id,
            "subject": course.subject.storageKey,
            "lesson_count": course.lessons.count,
            "has_quiz": course.hasQuiz,
        ])
    }

    static func trackCourseSessionEnded(
        courseId: String,
        subject: Subject,
        durationSeconds: Int,
        lessonsReached: Int,
        lessonCount: Int,
        continueTaps: Int,
        scrolledOnFirstLesson: Bool,
        engagementTier: String,
        exitReason: String
    ) {
        let progress = lessonCount > 0 ? Double(lessonsReached) / Double(lessonCount) : 0
        track("course_session_ended", [
            "course_id": courseId,
            "subject": subject.storageKey,
            "duration_seconds": durationSeconds,
            "lessons_reached": lessonsReached,
            "lesson_count": lessonCount,
            "progress_pct": progress,
            "continue_taps": continueTaps,
            "scrolled_on_first_lesson": scrolledOnFirstLesson,
            "engagement_tier": engagementTier,
            "exit_reason": exitReason,
        ])
    }

    // MARK: - Quiz

    static func trackQuizStarted(course: Course) {
        track("quiz_started", [
            "course_id": course.id,
            "subject": course.subject.storageKey,
            "question_count": course.quiz.count,
        ])
    }

    static func trackQuizCompleted(course: Course, score: Int, total: Int) {
        track("quiz_completed", [
            "course_id": course.id,
            "score": score,
            "total": total,
            "passed": score == total,
        ])
    }

    // MARK: - Streak & gates

    static func trackStreakUpdated(streakDays: Int, isNewRecord: Bool) {
        track("streak_updated", [
            "streak_days": streakDays,
            "is_new_record": isNewRecord,
        ])
        Mixpanel.mainInstance().people.set(property: "streak_current", to: streakDays)
    }

    static func trackFreemiumGateHit(
        gateType: String,
        subject: Subject?,
        courseId: String?
    ) {
        var props: [String: MixpanelType] = ["gate_type": gateType]
        if let subject { props["subject"] = subject.storageKey }
        if let courseId { props["course_id"] = courseId }
        track("freemium_gate_hit", props)
    }

    static func trackLockedContentTapped(subject: Subject, surface: String) {
        track("locked_content_tapped", [
            "subject": subject.storageKey,
            "surface": surface,
        ])
    }

    // MARK: - Deep links

    static func trackDeepLinkOpened(courseId: String) {
        track("deep_link_opened", ["course_id": courseId])
    }

    // MARK: - Feedback

    static func trackFeedbackOpened() {
        track("feedback_opened")
    }

    static func trackFeedbackSubmitted(category: String) {
        track("feedback_submitted", ["category": category])
    }

    static func trackFeedbackFailed(category: String) {
        track("feedback_failed", ["category": category])
    }

    // MARK: - Helpers

    static func onboardingStepName(for index: Int) -> String {
        switch index {
        case 0: "welcome"
        case 1: "language"
        case 2: "objective"
        case 3: "right_place"
        case 4: "swipe_courses"
        case 5: "phone_time"
        case 6: "weeks_lost"
        case 7: "review"
        case 8: "loading"
        case 9: "login"
        case 10: "trial_steps"
        case 11: "reminder"
        case 12: "paywall_annual"
        case 13: "paywall_comparison"
        default: "unknown"
        }
    }

    static func engagementTier(
        durationSeconds: Int,
        lessonsReached: Int,
        lessonCount: Int,
        continueTaps: Int,
        scrolledOnFirstLesson: Bool,
        completed: Bool
    ) -> String {
        if completed { return "completed" }
        let progress = lessonCount > 0 ? Double(lessonsReached) / Double(lessonCount) : 0
        if durationSeconds < 8, lessonsReached <= 1, continueTaps == 0 {
            return "bounce"
        }
        if durationSeconds < 30, lessonsReached <= 1, !scrolledOnFirstLesson, continueTaps == 0 {
            return "peek"
        }
        if progress >= 0.75 || (durationSeconds > 90 && continueTaps >= 2) {
            return "read"
        }
        if durationSeconds > 30 || scrolledOnFirstLesson || lessonsReached > 1 {
            return "skim"
        }
        return "peek"
    }

    private static func track(_ event: String, _ properties: [String: MixpanelType] = [:]) {
        Mixpanel.mainInstance().track(event: event, properties: properties)
    }
}

File: android/app/src/main/java/app/rork/sophia/data/AnalyticsService.kt
package app.rork.sophia.data

import android.content.Context
import app.rork.sophia.AppConfig
import app.rork.sophia.BuildConfig
import com.mixpanel.android.mpmetrics.MixpanelAPI
import org.json.JSONArray
import org.json.JSONObject

class AnalyticsService(context: Context) {
    private val prefs = context.getSharedPreferences("sophia_prefs", Context.MODE_PRIVATE)
    private val mixpanel: MixpanelAPI? = runCatching {
        MixpanelAPI.getInstance(context, AppConfig.MIXPANEL_TOKEN, true).also {
            it.setServerURL("https://api-eu.mixpanel.com")
            it.registerSuperProperties(
                JSONObject()
                    .put("platform", "android")
                    .put("app_version", BuildConfig.VERSION_NAME),
            )
        }
    }.getOrNull()

    fun updateContext(
        language: String,
        isPremium: Boolean,
        onboardingCompleted: Boolean,
        unlockedSubjects: Collection<String> = emptyList(),
    ) {
        runCatching {
            mixpanel?.registerSuperProperties(
                JSONObject()
                    .put("language", language)
                    .put("is_premium", isPremium)
                    .put("onboarding_completed", onboardingCompleted)
                    .put("unlocked_subjects", JSONArray(unlockedSubjects.toList())),
            )
            mixpanel?.people?.set("interests", JSONArray(unlockedSubjects.toList()))
        }
    }

    fun identify(userId: String) {
        runCatching {
            mixpanel?.identify(userId)
            mixpanel?.people?.identify(userId)
        }
    }

    fun track(event: String, props: Map<String, Any?> = emptyMap()) {
        val json = JSONObject()
        props.forEach { (k, v) ->
            when (v) {
                null -> Unit
                is Collection<*> -> json.put(k, JSONArray(v))
                else -> json.put(k, v)
            }
        }
        mixpanel?.track(event, json)
    }

    fun trackAppOpened(source: String = "cold") = track("app_opened", mapOf("source" to source))

    fun trackSessionIfNeeded(isPremium: Boolean) {
        val now = System.currentTimeMillis() / 1000.0
        val last = prefs.getFloat(KEY_LAST_SESSION, 0f).toDouble()
        if (now - last < 30 * 60) return
        val first = prefs.getFloat(KEY_FIRST_OPEN, 0f).toDouble().let {
            if (it <= 0) {
                prefs.edit().putFloat(KEY_FIRST_OPEN, now.toFloat()).apply()
                now
            } else it
        }
        prefs.edit().putFloat(KEY_LAST_SESSION, now.toFloat()).apply()
        val days = ((now - first) / 86400).toInt().coerceAtLeast(0)
        track("session_started", mapOf("days_since_install" to days, "is_premium" to isPremium))
    }

    fun trackOnboardingStarted() = track("onboarding_started")
    fun trackOnboardingStep(stepIndex: Int, stepName: String, action: String? = null) =
        track(
            "onboarding_step_viewed",
            buildMap {
                put("step_index", stepIndex)
                put("step_name", stepName)
                if (action != null) put("action", action)
            },
        )

    fun trackOnboardingCompleted(sawPaywall: Boolean, isPremium: Boolean) =
        track(
            "onboarding_completed",
            mapOf("saw_paywall" to sawPaywall, "is_premium_at_exit" to isPremium),
        )

    fun trackOnboardingInterestsSet(objectives: Collection<String>) =
        track("onboarding_interests_set", mapOf("objectives" to objectives.toList()))

    fun trackPaywallViewed(context: String, courseId: String? = null) =
        track(
            "paywall_viewed",
            buildMap {
                put("context", context)
                if (courseId != null) put("trigger_course_id", courseId)
            },
        )

    fun trackPaywallDismissed(context: String, durationSeconds: Int) =
        track("paywall_dismissed", mapOf("context" to context, "duration_seconds" to durationSeconds))

    fun trackDiscountOfferViewed(source: String) =
        track("discount_offer_viewed", mapOf("source" to source))

    fun trackPurchaseCompleted(context: String, offeringId: String? = null, packageId: String? = null) {
        track(
            "purchase_completed",
            buildMap {
                put("context", context)
                if (offeringId != null) put("offering_id", offeringId)
                if (packageId != null) put("package_id", packageId)
            },
        )
        mixpanel?.people?.set("is_premium", true)
    }

    fun trackCourseOpened(courseId: String, subject: String, source: String, isFreeUser: Boolean) =
        track(
            "course_opened",
            mapOf(
                "course_id" to courseId,
                "subject" to subject,
                "source" to source,
                "is_free_user" to isFreeUser,
            ),
        )

    fun trackCourseCompleted(courseId: String, subject: String, lessonCount: Int, hasQuiz: Boolean) =
        track(
            "course_completed",
            mapOf(
                "course_id" to courseId,
                "subject" to subject,
                "lesson_count" to lessonCount,
                "has_quiz" to hasQuiz,
            ),
        )

    fun trackQuizStarted(courseId: String, subject: String, questionCount: Int) =
        track(
            "quiz_started",
            mapOf("course_id" to courseId, "subject" to subject, "question_count" to questionCount),
        )

    fun trackQuizCompleted(courseId: String, score: Int, total: Int, passed: Boolean) =
        track(
            "quiz_completed",
            mapOf("course_id" to courseId, "score" to score, "total" to total, "passed" to passed),
        )

    fun trackStreakUpdated(streakDays: Int, isNewRecord: Boolean) {
        track("streak_updated", mapOf("streak_days" to streakDays, "is_new_record" to isNewRecord))
        mixpanel?.people?.set("streak_current", streakDays)
    }

    fun trackFreemiumGateHit(gateType: String, subject: String? = null, courseId: String? = null) =
        track(
            "freemium_gate_hit",
            buildMap {
                put("gate_type", gateType)
                if (subject != null) put("subject", subject)
                if (courseId != null) put("course_id", courseId)
            },
        )

    fun trackDeepLinkOpened(courseId: String) =
        track("deep_link_opened", mapOf("course_id" to courseId))

    fun trackFeedbackOpened() = track("feedback_opened")
    fun trackFeedbackSubmitted(category: String) =
        track("feedback_submitted", mapOf("category" to category))

    fun trackFeedbackFailed(category: String) =
        track("feedback_failed", mapOf("category" to category))

    fun trackCourseSessionEnded(props: Map<String, Any?>) = track("course_session_ended", props)

    fun trackLockedContentTapped(
        gateType: String,
        courseId: String?,
        subject: String?,
        surface: String = "course",
    ) = track(
        "locked_content_tapped",
        buildMap {
            put("gate_type", gateType)
            put("surface", surface)
            if (courseId != null) put("course_id", courseId)
            if (subject != null) put("subject", subject)
        },
    )

    fun trackAmbassadorOpened() = track("ambassador_opened")
    fun trackAmbassadorSubmitted() = track("ambassador_submitted")

    fun trackCollectionProgressed(collectionId: String, completed: Int, total: Int, isComplete: Boolean) =
        track(
            "collection_progressed",
            mapOf(
                "collection_id" to collectionId,
                "completed_count" to completed,
                "total_count" to total,
                "is_complete" to isComplete,
            ),
        )

    companion object {
        private const val KEY_LAST_SESSION = "sophia_analytics_last_session"
        private const val KEY_FIRST_OPEN = "sophia_analytics_first_open"
    }
}
